saas/rti/adapters/base.py

```
@@ -389,20 +388,123 @@ def create_symbolic_link(link_path: str, target_path: str, working_directory: st
     run_command(f"ln -sf {target_path} {link_path}")
 
 
+class JobRunner(Thread):
+    def __init__(self, owner: RTIProcessorAdapter, job_type: str, context: JobContext, wd_path: str):
+        super().__init__()
+        self._owner = owner
+        self._job_type = job_type
+        self._context = context
+        self._wd_path = wd_path
+
+    @property
+    def job(self) -> Job:
+        return self._context.job
+
+    def run(self):
+        try:
+            if self._job_type == 'new':
+                # perform pre-execute routine
+                self._owner.pre_execute(self._wd_path, self._context)
+
+                # instruct processor adapter to execute the job
+                self._owner.execute(self._wd_path, self._context)
+
+            elif self._job_type == 'resume':
+                pass
+
+            else:
+                raise SaaSRuntimeException(f"unexpected job type '{self._job_type}'")
+
+            # set the job state to RUNNING
+            self._context.state = JobStatus.State.RUNNING
+
+            # connect to the job and monitor its progress
+            self._owner.connect_and_monitor(self._context)
+
+            # perform post-execute routine
+            self._owner.post_execute(self._context.job.id)
+
+            # if we reach here the job is either running or cancelled
+            assert (self._context.state in [JobStatus.State.RUNNING, JobStatus.State.CANCELLED])
```

It is not recommended to use `assert` in actual code. It is mostly used for debugging and testing. I think it's better to throw a proper exception here that we can catch below.

---

Fixed it.